lnrpc/routerrpc/forward_interceptor.go

```diff
@@ -0,0 +1,253 @@
+		}
+	})
+}
+
+// rpcInterceptor is a helper struct that handles the lifecycle of an rpc
+// interceptor streaming session.
+// It is created when the stream opens and drains when the stream closes.
+// It should be used from the same go routine and not built for concurrent
+// access.
+type rpcInterceptor struct {
+
+	// holdForwards is a map of current hold forwards and their corresponding
+	// ForwardResolver.
+	holdForwards map[channeldb.CircuitKey]htlcswitch.InterceptedForward
+
+	// stream is the bidirectional stream
+	stream Router_HtlcInterceptorServer
+}
+
+// newRPCInterceptor creates a new rpcInterceptor.
+func newRPCInterceptor(stream Router_HtlcInterceptorServer) *rpcInterceptor {
+	return &rpcInterceptor{
+		stream: stream,
+		holdForwards: make(
+			map[channeldb.CircuitKey]htlcswitch.InterceptedForward),
+	}
+}
+
+// hold just store the intercepted forward.
+func (r *rpcInterceptor) hold(forward htlcswitch.InterceptedForward) {
+	r.holdForwards[forward.CircuitKey()] = forward
+}
+
+// forwardToClient forwards the intercepted htlc to the client.
+func (r *rpcInterceptor) forwardToClient(forward htlcswitch.InterceptedForward) error {
+	htlc := forward.Packet()
+	inKey := forward.CircuitKey()
+
+	// A ChannelAcceptRequest has been received, send it to the client.
+	interceptionRequest := &ForwardHtlcInterceptRequest{
+		CircuitKey: &CircuitKey{
+			IncomingChanId: inKey.ChanID.ToUint64(),
+			HtlcId:         inKey.HtlcID,
+		},
+		HtlcPaymentHash: htlc.PaymentHash[:],
+		AmountSat:       uint64(htlc.Amount.ToSatoshis()),
+		Expiry:          htlc.Expiry,
+	}
+
+	return r.stream.Send(interceptionRequest)
+}
+
+// resolveFromClient handles a resolution arrived from the client.
+func (r *rpcInterceptor) resolveFromClient(in *ForwardHtlcInterceptResponse) error {
+	circuitKey := channeldb.CircuitKey{
+		ChanID: lnwire.NewShortChanIDFromInt(in.CircuitKey.IncomingChanId),
+		HtlcID: in.CircuitKey.HtlcId,
+	}
+	var interceptedForward htlcswitch.InterceptedForward
+	interceptedForward, ok := r.holdForwards[circuitKey]
+	if !ok {
+		return ErrFwdNotExists
+	}
+	delete(r.holdForwards, circuitKey)
+
+	switch in.Action {
+	case ResolveHoldForwardAction_RESUME:
+		return interceptedForward.Resume()
+	case ResolveHoldForwardAction_FAIL:
+		return interceptedForward.Fail()
+	case ResolveHoldForwardAction_SETTLE:
+		if in.RPreimage == nil {
+			return ErrMissingPreimage
+		}
+		preimage, err := lntypes.MakePreimage(in.RPreimage)
+		if err != nil {
+			return err
+		}
+		err = interceptedForward.Settle(preimage)
+		return err
+	default:
+		return fmt.Errorf("unrecognized resolve action %v", in.Action)
+	}
+}
+
+// drain removes all previousely held forwards and allows to pass a callback
+// that will be called for each one before it is removed.
+func (r *rpcInterceptor) drain(
+	resolver func(forward htlcswitch.InterceptedForward)) {
+
+	for key, forward := range r.holdForwards {
+		resolver(forward)
```

Comment:
Why can't this just be `forward.Resume()`?

Reply:
Left over from a previous iteration. Simplified now.